integration-tests/src/tests/client/process_blocks.rs

```diff
@@ -3900,6 +3902,265 @@ mod access_key_nonce_range_tests {
         assert_eq!(num_requests, 4 * 20);
     }
 
+    struct ChunkForwardingOptimizationTestData {
+        num_validators: usize,
+        env: TestEnv,
+
+        num_part_ords_requested: usize,
+        num_part_ords_sent_as_partial_encoded_chunk: usize,
+        num_part_ords_forwarded: usize,
+        num_forwards_with_missing_chunk_header: usize,
+        chunk_parts_that_must_be_known: HashSet<(ChunkHash, u64, usize)>,
+    }
+
+    impl ChunkForwardingOptimizationTestData {
+        fn new() -> ChunkForwardingOptimizationTestData {
+            let num_clients = 4;
+            let num_validators = 4 as usize;
+            let num_block_producers = 1;
```

Comment:
Is your intended setup 1 block producer plus three chunk only producers? Did you check if that's indeed the case?

Reply:
I did check by printing out the epoch info - and in fact if this weren't true the test would not pass. If there were 4 block producers, then client 0 cannot successfully produce a block every time; on the other hand if there were 1 chunk producer, no forwarding would ever be performed.